abaco.go

```diff
@@ -86,138 +318,79 @@ func (device *AbacoDevice) ReadAllPackets() ([]*packets.Packet, error) {
 	return allPackets, nil
 }
 
-// sampleCard samples the data from a single card to scan enough packets to
-// know the number of channels, data rate, etc.
-// Although it slows things down, it's best to discard all data in the ring
-// at the time we open it, because we have no idea how old the data are.
-func (device *AbacoDevice) sampleCard() error {
-	// Open the device and discard whatever is in the buffer
-	if err := device.ring.Open(); err != nil {
-		return err
+// samplePackets samples the data from a single AbacoRing. It scans enough packets to
+// learn the number of channels, data rate, etc.
+// Although it slows things down, it's best to discard all data in the ring when
+// we open it, because we have no idea how old the data are (in particular, once full, a
+// ring cannot take in new data to replace the old, so a full ring's data are arbitrarily old).
+// TODO: consider whether the newest data a *non-full* ring can be used here?
+func (device *AbacoRing) samplePackets() (allPackets []*packets.Packet, err error) {
+	// Open the ring buffer and discard whatever is in it.
+	if err = device.ring.Open(); err != nil {
+		return
 	}
 	psize, err := device.ring.PacketSize()
 	if err != nil {
-		return err
+		return
 	}
 	device.packetSize = int(psize)
-	if err := device.ring.DiscardStride(uint64(device.packetSize)); err != nil {
-		return err
+	if err = device.ring.DiscardStride(uint64(device.packetSize)); err != nil {
+		return
 	}
 
-	// Now get the data we actually want. Run for at least a minimum time
-	// or a minimum number of packets.
-	const minPacketsToRead = 100 // Not sure this is a good minimum
-	maxDelay := time.Duration(2000 * time.Millisecond)
-	timeOut := time.NewTimer(maxDelay)
-
-	// Capture timestamp and sample # for a range of packets. Use to find rate.
-	var tsInit, tsFinal packets.PacketTimestamp
-	var snInit, snFinal uint32
-	samplesInPackets := 0
+	// Now get the data we actually want: fresh data. Run for at least
+	// a minimum time or a minimum number of packets.
+	const minPacketsToRead = 100 // Not sure this is a good minimum?
```

Comment:
what is needed for success? one from each channel group? would be nice to have a comment describing what is neccesary to learn all you need

Reply:
Good point. I will attempt to spell it out in a comment, though I'm not sure I really know what's needed. We need "some data", with at least 2 packets. That's because we need the time between packets (as time-stamped by the firmware on the Abaco) to estimate sample rates.